src/coreComponents/finiteElement/elementFormulations/FiniteElementBase.hpp

```diff
@@ -13,13 +35,49 @@ namespace finiteElement
 {
 
 /**
- * @class FiniteElementShapeFunctionKernelBase
- * @brief Base class for the finite element kernels.
+ * @brief Base class for FEM element implementations.
  */
 class FiniteElementBase
 {
 public:
 
+  /// Default Constructor
+  FiniteElementBase() = default;
+
+  /**
+   * @brief Copy Constructor
+   * @param source The object to copy.
+   */
+  FiniteElementBase( FiniteElementBase const & source ):
+#ifdef CALC_FEM_SHAPE_IN_KERNEL
```

Comment:
I was having a problem getting rid of the preprocessor completely. The problem is that we have these member views (`m_viewGradN, m_viewDetJ`), and these are set on the host. For host execution it seems better to use the precomputed values...so great. However, for device execution, we certainly don't want to use these values, NOR do we want to move the views to the device, which is what happens in the default CC.

I would like to get rid of the setters, but there was some reordering in the initialization that needs to be done. I should just do that with this PR...but I had forgot. I'll do that before merging.

Reply:
Ah yeah if you took a `FiniteElementBase` which had the views set then the default copy constructor isn't what you want. But if they weren't set then it's fine.

RAJA does have a trait for that, I don't remember what it is off the top of my head.

Doing away with the preprocessor entirely would either require some templating or divergent behavior in host and device code. I don't think we want the divergent host-device behavior because there are times when you want to calculate the shape functions on host (like in initialization).

Reply:
Well in the case of initialization we can just call the `calcShapeFuncitonDerivatives()` directly. Which reminds me...I need to change that name to `calcGradN`